### DISCUSSION TITLE
Problem with the memory

### DISCUSSION THREAD

**Discussion by parham-s-96** [pelicanmapping/osgearth#2674] [Support Q&A]
Hello guys.
I have a problem with this code : 

        const ref_ptr<Group> tileNodes = new Group();
        osgUtil::IntersectionVisitor iVisitor;
        osg::ref_ptr<osgUtil::LineSegmentIntersector>line = new osgUtil::LineSegmentIntersector(vec1, vec2);
        iVisitor.setIntersector(line);

        for(double i = 0.0 ; i <= 45.0 ; i += 0.001)
        {
            const GeoPoint point = GeoPoint(mapNode->getMapSRS() , i , i , AltitudeMode::ALTMODE_ABSOLUTE);
            const TileKey targetTileKey = mapNode->getMap()->getProfile()->createTileKey(point.x(), point.y(), 16u);

            QSet<TileKey> collectedTiles;
            if (collectedTiles.contains(targetTileKey) == false)
            {
                const osg::ref_ptr<osg::Node> tileNode = mapNode->getTerrainEngine()->createTile(targetTileKey);
                const osg::ref_ptr<Drivers::MPTerrainEngine::TileNode> createdTileGeometryNode = osg::dynamic_pointer_cast<Drivers::MPTerrainEngine::TileNode*>(tileNode);

                if (createdTileGeometryNode->isValid())
                {
                    collectedTiles.insert(targetTileKey);
                    tileNodes->addChild(createdTileGeometryNode);
                }
            }
        }

        tileNodes->accept(iVisitor);
        osgUtil::LineSegmentIntersector::Intersections &hits = line->getIntersections();

When I run it, the memory gets filled up a lot and is not freed. At first I thought it was because of a memory leak. But when I looked further, I realized that this happens the first time I run this code in my program and when I run this code again after the first run, no more memory is filled.
I guess it's something like a caching system, but anyway, it's causing problems because a lot of memory stays filled up until I close the program.
Thanks for your help.

**Comment by gwaldron** [comment]
You didn't say what version you are using. Looks very old. Maybe you can explain what exactly you are trying to accomplish?

Also, moving `collectedTiles` outside the `for` loop might help.

**Comment by parham-s-96** [comment]
the osgEarth version is 2.9

here is my code snippet:
about what i am trying to accomplish here : i collect tiles for a prespecified radius(trying to avoid duplicate tiles)
gathering them into a groupNode then intersect them with bunch of lineSegmentIntersectors.
As i mentioned before, the memory gets filled up and it is not freed until I close the application.


for (size_t i = 0; i < iGroupSize; i++)
    {
      if (_terminate) { break; }

      auto  ls = dynamic_cast<osgUtil::LineSegmentIntersector *>(_ivGroup->getIntersectors()[i].get());
      pointL2.fromWorld(mapNode->getMapSRS(), ls->getEnd());

      double  tScale    = 0.0;
      size_t  increment = 1;
      tScale = static_cast<double>(increment * tileXLength / r);  // r : radius(prespecified), tileXLength : 245.0 meter(double variable)

      while (1 >= tScale)
      {
      // lerp() : get a full length line(start, end), outputs the end point of of scaled line based on the scale param. 0<scale param<1
      // for a example : pointL1(0,0,0)  pointL2(10,0,0)  scale param = 0.1  result = p(1,0,0)
        auto  resultLerpPoint = GeoUtils::geoLerp(pointL1, pointL2, tScale);
        
        auto  targetTileKey   = mapNode->getMap()->getProfile()->createTileKey(resultLerpPoint.x(), resultLerpPoint.y(), 16u);

        if (!_collectedTileKeys->contains(targetTileKey))  // checkking if the i have already collected this tile(by tile key)
        {
          auto  tNode = mapNode->getTerrainEngine()->createTile(targetTileKey);

          if (nullptr != tNode)
          {
            auto  createdTileGeometryNode = dynamic_cast<osgEarth::Drivers::MPTerrainEngine::TileNode *>(tNode);

            if (nullptr != createdTileGeometryNode)
            {
              if (createdTileGeometryNode->isValid())
              {
                _collectedTileKeys->insert(targetTileKey);
                out_osgGroupTileNodes->addChild(createdTileGeometryNode)
              }
            }
          }
        }

        ++increment;
        tScale = static_cast<double>(increment * tileXLength / r);
      }

**Reply by gwaldron** [reply]
I'll give you a couple notes, but there is not much I can do if there's a problem in osgEarth 2.9 itself. Too old. Anyway:
- If any of your "if" statements fails, createdTileGeometryNode will leak memory since you are not using a ref_ptr.
- Even if your tile is already in _collectedTileKeys, you are still adding it to your group. So you are likely adding the same tiles over and over.
- Can't tell from this code, but make sure out_osgGroupTileNodes is in a ref_ptr too so it can be properly deleted.

Good luck.
